Propagating a Proxy Config [JSON] String from browser down to go-tun2socks layer

bi-kh opened the pull request:
The config string is mostly opaque (except host retrieval in iOS) and passed along the layers unchanged.

Note: In iOS, We no longer update the config's host with the resolved IP address and pass along the original hostname, if not an IP already, to tun2socks layer so the config string effectively remains unchanged all the way.

bi-kh committed:
Propagating a Proxy Config [JSON] String from browser down to go-tun2socks layer

The config string is mostly opaque (except host retrieval in iOS) and passed along the layers unchanged.

Note: In iOS, We no longer update the config's host with the resolved IP address and pass along the original hostname, if not an IP already, to tun2socks layer so the config string effectively remains unchanged all the way.

fortuna commented on src/cordova/android/OutlineAndroidLib/outline/src/main/aidl/org/outline/TunnelConfig.aidl:
```diff
@@ -14,10 +14,7 @@
   String id;
-  String name;
-  ShadowsocksConfig proxy;
+  String proxyString;  
```
The concepts of a service configuration in the app is a different concept from the networking configuration. We must represent them differently.

The name should stay in the TunnelConfig, not in the proxyString. The proxy string should be for configuring the VPN only, and the name is not needed for networking.

bi-kh replied:
'name' was not set before as far as I see. It seemed redundant and removed.

fortuna commented on src/cordova/android/OutlineAndroidLib/outline/src/main/java/org/outline/vpn/VpnTunnelService.java:
```diff
@@ -199,46 +199,19 @@ public VpnService.Builder newBuilder() {
    * @return populated TunnelConfig
    */
-  public static TunnelConfig makeTunnelConfig(final String tunnelId, final JSONObject config)
+  public static TunnelConfig makeTunnelConfig(final String tunnelId, final String configString)
```
Revert to the JSONObject. You'll extract the proxy string from a field in the JSON.

bi-kh replied:
See _CordovaTunnel_ in cordova_main.ts
_start_ receives a  _ShadowsocksSessionConfig_, right? And now I pass the opaque string version rather than its original JSON format. 
Do we intend to introduce a new container config which wraps this proxy config [string] and pass that from Cordova layer?

fortuna replied:
I believe we can get rid of the IS_REACHABLE call too. It's not really useful. We use it to determine the status of the Tunnel, but we can do that in a different way. It makes sense to have a separate PR to just remove IsReachable first.

fortuna commented on src/cordova/android/OutlineAndroidLib/outline/src/main/java/org/outline/vpn/VpnTunnelService.java:
```diff
@@ -341,7 +308,7 @@ private synchronized boolean isTunnelActive(final String tunnelId) {
 
   private boolean isServerReachable(final String host, final int port) {
```
I don't think this is needed, and it forces the app to dig into the config. We should get rid of it.

fortuna commented on src/cordova/android/OutlineAndroidLib/outline/src/main/java/org/outline/vpn/VpnTunnelService.java:
```diff
@@ -621,17 +573,23 @@ public int getResourceId(final String name, final String type) {
-  /* Returns the server's name from |serverConfig|. If the name is not present, it falls back to the
+  /* Returns the server's name from |proxyString|. If the name is not present, it falls back to the
    * host name (IP address), or the application name if neither can be retrieved. */
   private String getServerName(final TunnelConfig config) {
```
Revert

fortuna replied:
Qualify the config.
```suggestion
  private String getServerName(final TunnelConfig tunnelConfig) {
```

fortuna commented on src/cordova/apple/OutlineAppleLib/Sources/OutlineTunnelSources/OutlineTunnel.swift:
```diff
@@ -14,25 +14,10 @@
-    return ["host": host ?? "", "port": port ?? "", "password": password ?? "",
-            "method": method ?? "", "prefix": prefixStr]
-  }
+  public var configString: String?
```
Name it proxyString or proxyJson, so it's clearer.

fortuna replied:
Qualify the config:
```suggestion
  public var proxyConfigString: String?
```

fortuna commented on src/cordova/apple/OutlineAppleLib/Sources/OutlineTunnelSources/OutlineTunnel.swift:
```diff
@@ -41,26 +26,39 @@ public class OutlineTunnel: NSObject, Codable {
-  public static func decode(_ jsonData: Data) -> OutlineTunnel? {
-    return try? JSONDecoder().decode(OutlineTunnel.self, from: jsonData)
+  // Private helper to retrieve the host from the config string.
+  private func configToDictionary() -> [String: Any]? {
```
This means the config is not opaque to the platform-specific code, defeating the purpose of wiring the config.

What is the host used for besides the reachability test?

We may need to think about what we will show in the case of dynamic keys and what abstractions we need to support that. Perhaps we need to give different configs different ids. Or have the `Client` have a method to return the information to show.

fortuna replied:
The action item here is to first understand what the host is being used for. We've talked about the Is Reachable: we should get rid of that. What else?

fortuna commented:
FYI, we've finally merged the change to support config-based clients. See https://github.com/Jigsaw-Code/outline-go-tun2socks/pull/117

That means some of the calls will need to change.

fortuna commented on src/cordova/android/OutlineAndroidLib/outline/src/main/java/org/outline/vpn/VpnTunnelService.java:
```diff
@@ -199,46 +199,19 @@ public VpnService.Builder newBuilder() {
    * @return populated TunnelConfig
    */
-  public static TunnelConfig makeTunnelConfig(final String tunnelId, final JSONObject config)
+  public static TunnelConfig makeTunnelConfig(final String tunnelId, final String configString)
```
Let's qualify all mentions to "config", since it's ambiguous.
```suggestion
  public static TunnelConfig makeTunnelConfig(final String tunnelId, final String proxyConfigString)
```

fortuna commented on src/cordova/android/OutlineAndroidLib/outline/src/main/java/org/outline/vpn/VpnTunnelService.java:
```diff
@@ -251,7 +224,7 @@ private ErrorCode startTunnel(final TunnelConfig config) {
       final TunnelConfig config, boolean isAutoStart) {
     LOG.info(String.format(Locale.ROOT, "Starting tunnel %s.", config.id));
-    if (config.id == null || config.proxy == null) {
+    if (config.id == null || config.proxyString == null) {
```
```suggestion
    if (tunnelConfig.id == null || tunnelConfig.proxyConfigString == null) {
```

fortuna commented on src/cordova/android/OutlineAndroidLib/outline/src/main/java/org/outline/vpn/VpnTunnelService.java:
```diff
@@ -621,17 +573,23 @@ public int getResourceId(final String name, final String type) {
-      String serverName = config.name;
-      if (serverName == null || serverName.equals("")) {
-        serverName = config.proxy.host;
+      JSONObject object = (JSONObject) new JSONTokener(config.proxyString).nextValue();
```
We should treat the proxy config as opaque. Move the name back to the TunnelConfig.

fortuna commented on src/cordova/apple/OutlineAppleLib/Sources/OutlineTunnelSources/OutlineTunnel.swift:
```diff
@@ -41,26 +26,39 @@ public class OutlineTunnel: NSObject, Codable {
   }
 
-  public convenience init(id: String, config: [String: Any]) {
+  public convenience init(id: String?, configString: String) {
```
```suggestion
  public convenience init(id: String?, proxyConfigString: String) {
```

fortuna commented on src/cordova/plugin/android/java/org/outline/OutlinePlugin.java:
```diff
@@ -191,8 +191,8 @@ private void executeAsync(
-          final JSONObject config = args.getJSONObject(1);
-          int errorCode = startVpnTunnel(tunnelId, config);
+          final String configString = args.getString(1);
+          int errorCode = startVpnTunnel(tunnelId, configString);
```
Qualify the config

fortuna commented on src/cordova/plugin/android/java/org/outline/OutlinePlugin.java:
```diff
@@ -259,11 +259,11 @@ public void onActivityResult(int request, int result, Intent data) {
   }
 
-  private int startVpnTunnel(final String tunnelId, final JSONObject config) throws Exception {
+  private int startVpnTunnel(final String tunnelId, final String configString) throws Exception {
```
Qualify the config.

fortuna commented on src/cordova/plugin/apple/src/OutlinePlugin.swift:
```diff
@@ -76,11 +76,11 @@ class OutlinePlugin: CDVPlugin {
                        errorCode: OutlineVpn.ErrorCode.illegalServerConfiguration)
     }
-    let tunnel = OutlineTunnel(id: tunnelId, config: config)
+    let tunnel = OutlineTunnel(id: tunnelId, configString: configString)
```
Qualify the config.

bi-kh committed:
Update (#2) on propagating Proxy Config String from the browser layer

- revises Cordova 'start' command arguments# to 1 by wrapping (id, host, proxyConfigstring) in TunnelConfig. This will simplify future updates.
- propagates 'host' explicitly to keep proxyConfigString opaque
- simplifies some method arguments
- revises OutlineTunnel.swift

Note: not tested yet

fortuna commented on src/cordova/plugin/apple/src/OutlinePlugin.swift:
```diff
@@ -67,20 +67,18 @@ class OutlinePlugin: CDVPlugin {
   func start(_ command: CDVInvokedUrlCommand) {
-    guard let tunnelId = command.argument(at: 0) as? String else {
-      return sendError("Missing tunnel ID", callbackId: command.callbackId,
+    guard let tunnelCnfig = command.argument(at: 0) as? [String: Any] else {
```
Cnfig -> Config

fortuna reviewed (commented):
FYI, outline-go-tun2socks is not ready yet. We are [working on a fix](https://github.com/Jigsaw-Code/outline-go-tun2socks/pull/118), then we'll release the library to use in outline-client.

bi-kh committed:
Update #3 on propagating Proxy Config String from the browser layer  (iOS)

- Refactors app/VpnExtension layers 'start' args to a 'Data" encoded 'OutlineTunnel' format instead of individual arg for each property
- Compiled/tested iOS

TODO:
- Compile/test Android
- Update the app/tun2socks interface from the initial ProxyClient to the merged tun2socks

fortuna commented on src/cordova/android/OutlineAndroidLib/outline/src/main/java/org/outline/vpn/VpnTunnel.java:
```diff
@@ -119,19 +119,19 @@ public synchronized void tearDownVpn() {
+   * Connects a tunnel between a proxy server and the VPN TUN interface.
    *
-   * @param client provides access to the Shadowsocks proxy.
+   * @param client provides access to the  proxy.
```
```suggestion
   * @param client provides access to the proxy.
```

fortuna commented on src/cordova/android/OutlineAndroidLib/outline/src/main/java/org/outline/vpn/VpnTunnelService.java:
```diff
@@ -251,7 +227,7 @@ private ErrorCode startTunnel(final TunnelConfig config) {
   private synchronized ErrorCode startTunnel(
       final TunnelConfig config, boolean isAutoStart) {
```
I'd like to reduce the dependency on the TunnelConfig and be more explicit about what's needed here:
```suggestion
  private synchronized ErrorCode startTunnel(
      String id, String proxyConfigString, boolean isAutoStart) {
```

In fact, can we get rid of TunnelConfig and makeTunnelConfig for now?

fortuna commented on src/cordova/plugin/android/java/org/outline/OutlinePlugin.java:
```diff
@@ -259,13 +258,14 @@ public void onActivityResult(int request, int result, Intent data) {
 
-  private int startVpnTunnel(final String tunnelId, final JSONObject config) throws Exception {
-    LOG.info(String.format(Locale.ROOT, "Starting VPN tunnel %s", tunnelId));
+  private int startVpnTunnel(final JSONObject jsonTunnelConfig) throws Exception {
```
Let's revert to the more explicit interface we had before. 
I'd rather not have the TunnelConfig for now, as we don't configure anything else other than the proxy.
Also, I'm not sure the id belongs in a tunnel config if it's not used to configure the VPN.

fortuna commented on src/cordova/android/OutlineAndroidLib/outline/src/main/java/org/outline/vpn/VpnTunnelService.java:
```diff
@@ -621,19 +575,9 @@ public int getResourceId(final String name, final String type) {
-  /* Returns the server's name from |serverConfig|. If the name is not present, it falls back to the
-   * host name (IP address), or the application name if neither can be retrieved. */
+  /* Returns the server's host name (IP address). */
   private String getServerName(final TunnelConfig config) {
```
It looks like this is only used [to build the foreground service notification](https://github.com/Jigsaw-Code/outline-client/blob/3c015bde60d0142b8dc39707aa4ac72f49632ede/src/cordova/android/OutlineAndroidLib/outline/src/main/java/org/outline/vpn/VpnTunnelService.java#L595).

We should be more explicit about that. If we need a name to identify the service we are using, we should pass that down instead of a host address. The name is nice because it's protocol-agnostic. Some strategies may need multiple IP addresses, which doesn't fit the host address model.

It seems we need:
- id (why?)
- service name (for system VPN UI)
- proxy config (for setting up the proxy)

fortuna commented on src/cordova/android/OutlineAndroidLib/outline/src/main/aidl/org/outline/TunnelConfig.aidl:
```diff
@@ -14,10 +14,8 @@
   String id;
-  String name;
-  ShadowsocksConfig proxy;
+  String host;
```
Revert. We need a service name, not a host address.
Perhaps we should rename the field to serviceName.

fortuna commented on src/cordova/android/OutlineAndroidLib/outline/src/main/aidl/org/outline/TunnelConfig.aidl:
```diff
@@ -14,10 +14,8 @@
 
-import org.outline.shadowsocks.ShadowsocksConfig;
-
 parcelable TunnelConfig {
```
I'm, not convinced we need this type, versus passing the values that are needed explicitly in the different points. The extra type adds an intermediate step that sometimes hurts readability, especially since not all fields are needed everywhere this type is being used.

bi-kh committed:
Update 4 on propagating Proxy Config String from the browser layer (Android)

- Adds 'serverName' to the browser layer TunnelConfig and fills it with 'host' for now
- Renames Android TunnelConfig's 'host' to 'serverName'
- Tested

TODO:
- Update the app/tun2socks interface from the initial ProxyClient to the merged tun2socks (iOS & Android)
- Wire the actual 'serverName' to the browser's TunnelConfig

bi-kh committed:
Update 5 on propagating Proxy Config String from the browser layer (Android)

- Wired the server name to the browser-side TunnelConfig propagated to the native-side 'start'. Verified that Android makeTunnelConfig retrieves it properly.

TODO:
- Update the app/tun2socks interface from the initial ProxyClient to the merged tun2socks (iOS & Android)
